src/data/feature_index.js

```diff
@@ -125,79 +114,112 @@ class FeatureIndex {
 
         const matching = this.grid.query(minX - queryPadding, minY - queryPadding, maxX + queryPadding, maxY + queryPadding);
         matching.sort(topDownFeatureComparator);
-        this.filterMatching(result, matching, this.featureIndexArray, queryGeometry, filter, params.layers, styleLayers, pixelsToTileUnits, args.posMatrix, args.transform);
+        const result = {};
+        let previousIndex;
+        for (let k = 0; k < matching.length; k++) {
+            const index = matching[k];
+
+            // don't check the same feature more than once
+            if (index === previousIndex) continue;
+            previousIndex = index;
 
-        const matchingSymbols = args.collisionIndex ?
-            args.collisionIndex.queryRenderedSymbols(queryGeometry, this.tileID, args.tileSize / EXTENT, args.collisionBoxArray, args.sourceID, args.bucketInstanceIds) :
-            [];
-        matchingSymbols.sort();
-        this.filterMatching(result, matchingSymbols, args.collisionBoxArray, queryGeometry, filter, params.layers, styleLayers, pixelsToTileUnits, args.posMatrix, args.transform);
+            const match = this.featureIndexArray.get(index);
+            let geometry = null;
+            this.generateGeoJSONFeature(
+                result,
+                match.bucketIndex,
+                match.sourceLayerIndex,
+                match.featureIndex,
+                filter,
+                params.layers,
+                styleLayers,
+                (feature: VectorTileFeature, styleLayer: StyleLayer) => {
+                    if (!geometry) {
+                        geometry = loadGeometry(feature);
+                    }
+                    return styleLayer.queryIntersectsFeature(queryGeometry, feature, geometry, this.z, args.transform, pixelsToTileUnits, args.posMatrix);
+                }
+            );
+        }
 
         return result;
     }
 
-    filterMatching(
+    generateGeoJSONFeature(
```

Comment:
Naming nit: feels to me like the most important thing this method is doing is still deciding whether or not to include the feature, moreso than generating the geojson result if the answer is 'yes'.  Bad suggested rename: `maybeAddFeatureToResult()`

Reply:
🤔You're right, when I first named this I thought the two pathways would end up sharing a smaller chunk of shared code. Maybe I should just go back to `filterMatching`? Or maybe `loadMatchingFeatures`?

Reply:
Went with `loadMatchingFeature`